### PR TITLE
Import bhmm to support hidden markov state models

### PR DESCRIPTION
**Pull request by marscher**
TODO:

- [ make HMM.submodel independent of dtrajs, states and obs arguments should only accept int array or None, provide methods to compute these arrays.]

### CODE REVIEW

File: sktime/markovprocess/bhmm/hidden/src/bhmm_hidden_module.cpp [hunk 1]
```diff
@@ -263,31 +244,28 @@ np_array<std::int32_t> samplePath(const np_array<dtype> &alpha, const np_array<d
...
+        path[T - 1] = ddist(generator); //_random_choice(psel, N);
 
         // Work backwards from T-2 to 0.
         for (std::size_t t = T - 1; t >= 1; t--) {
             // Compute P(s_t = i | s_{t+1}..s_T).
             for (std::size_t i = 0; i < N; i++) {
                 psel[i] = alphaBuf[(t - 1) * N + i] * ABuf[i * N + path[t]];
             }
-            normalize(psel, psel + N);
+            std::discrete_distribution<> ddistBackw (psel, psel+N);
```
Comment (marscher) [inline sktime/markovprocess/bhmm/hidden/src/bhmm_hidden_module.cpp]: is it mandatory to recreate this object in the loop or is there a way to re-initialize it?

Comment (clonker) [reply sktime/markovprocess/bhmm/hidden/src/bhmm_hidden_module.cpp]: there is a way to re-initialize it!

Comment (marscher) [reply sktime/markovprocess/bhmm/hidden/src/bhmm_hidden_module.cpp]: muy bien!

File: sktime/markovprocess/bhmm/hmm/generic_hmm.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+        if not self.hidden_state_trajectories:
+            raise RuntimeError('HMM model does not have a hidden state trajectory.')
+        dtype = observations[0].dtype
+        collected_observations = np.array([], dtype=dtype)
+        for (s_t, o_t) in zip(self.hidden_state_trajectories, observations):
+            indices = np.where(s_t == state_index)[0]
+            collected_observations = np.append(collected_observations, o_t[indices])
+
+        return collected_observations
+        collected_observations = (o_t[np.where(s_t == state_index)[0]]
```
Comment (clonker) [inline sktime/markovprocess/bhmm/hmm/generic_hmm.py]: this is after the return statement

**Comment by clonker** [comment]
I am merging this now as this was mostly meant to import bhmm into the source tree. Needs subsequent cleanup works and in particular the CK test for hmms is unclear.

**Pull request merged**